### PR TITLE
Type checking 'for...of' in ES3/5

### PR DESCRIPTION
**Pull request by JsonFreeman**
A 'for...of' statement in ES5 can take an array, a string, a union of arrays, or a union of arrays and string.

A 'for...of' statement in ES3 can take an array or a union of arrays.

### CODE REVIEW

**Comment by JsonFreeman** [comment]
My last commit is entitled "Rebaseline tests that will be affected by #2308". I meant #2309.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8777,16 +8781,11 @@ module ts {
         }
 
         function checkForOfStatement(node: ForOfStatement): void {
-            if (languageVersion < ScriptTarget.ES6) {
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: Nice!

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -8987,6 +8978,45 @@ module ts {
             }
         }
 
+        function checkElementTypeOfArrayOrString(arrayOrStringType: Type, expressionForError: Expression): Type {
+            Debug.assert(languageVersion < ScriptTarget.ES6);
+            var isJustString = allConstituentTypesHaveKind(arrayOrStringType, TypeFlags.StringLike);
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: isNotUnjustString

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -8987,6 +8978,45 @@ module ts {
             }
         }
 
+        function checkElementTypeOfArrayOrString(arrayOrStringType: Type, expressionForError: Expression): Type {
+            Debug.assert(languageVersion < ScriptTarget.ES6);
+            var isJustString = allConstituentTypesHaveKind(arrayOrStringType, TypeFlags.StringLike);
+            
+            // Check isJustString because removeTypesFromUnionType will only remove types if it doesn't result
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: I feel like this behavior should be in removeTypesFromUnionType.  i.e. have a flag like 'allowAllTypesToBeRemoved'.   It's very weird that 'removeTypesFromUnionType' will choose to not actually remove the type, and that you have to work around that here.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -8987,6 +8978,45 @@ module ts {
...
+                reportedError = true;
+            }
+
+            if (isJustString) {
+                return stringType;
+            }
+
+            if (!isArrayLikeType(arrayType)) {
+                if (!reportedError) {
+                    error(expressionForError, Diagnostics.Type_0_is_not_an_array_type, typeToString(arrayType));
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: SHould this be 'is not an array or string type'?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -8987,6 +8978,45 @@ module ts {
...
+                }
+                return hasStringConstituent ? stringType : unknownType;
+            }
+
+            var arrayElementType = getIndexTypeOfType(arrayType, IndexKind.Number) || unknownType;
+            if (hasStringConstituent) {
+                // This is just an optimization for the case where arrayOrStringType is string | string[]
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: Could you add, at the top of this function, the types of types you want this fnction to be able to handle, and what final type will be returned?

Note: this function never seems to return 'undefined'.  But i thought you used || earlier to default to hte 'any' type of you got 'undefined' back from this.

File: src/compiler/checker.ts (hunk 3 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Yeah I think you are right. I was actually thinking about that when I did this. I can add a parameter called allowEmptyUnionResult.

File: src/compiler/checker.ts (hunk 4 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Maybe... but if you give the type `string | number`, it will say that type `number` is not an array type.

Comment (JsonFreeman) [inline src/compiler/checker.ts]: Yet, if you give the type `number`, it should say that `number` is not a string or array type.

File: src/compiler/checker.ts (hunk 5 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Yeah I will add a summary of what the result should be, as well as some examples.

**Comment by yuit** [comment]
:+1:

**Pull request merged**